Fix #7, add page about Shares

paldys opened the pull request:
After this change a new page will be added, where one can look at their shares.

Issues addressed in this PR, which were necessary / desired to add a new page to pay-overview:
- **Fix ordering of items by date** - bug fix
- **Drop the usage of 'parse' from 'date-fns'** - refactoring, no need for parse
- **Use the same date/amount formatting across the project** - refactoring, code duplication
- **Move column width sizes to the rest of the column data** - refactoring, let's keep similar things together
- **Generalize OverviewTable, move specifics to OverviewPage** - OverviewTable had information about the data it uses, it will be passed as parameter
- **Rename Overview -> Payroll** - let's specify what is related to Payroll
- **Extract API call to a separate file** - refactor, create a file for API calls
- **Throw an error on unsuccessful API calls** - enhance API calls with errors
- **Introduce a Fetch component** - a component to deal with API calls
- **Show error message on failed request** - show errors if API calls fail and do not silently ignore them
- **Extract DateFilter as a separate component** - a reusable component for date filtering
- **Introduce a common Layout component** - a reusable component for Layout
- **Introduce navigation using React Router** - navigation through pages and 'Page not found'
- **Refactor server/api.js** - create reusable methods for accessing google sheet data
- **Fix #7, add page about Shares** - introduce the new page.

paldys commented on client/src/utils/format.js:
```diff
@@ -0,0 +1,17 @@
+  year: 'numeric',
+  month: 'short',
+  day: '2-digit',
+})
```
This is a private helper... and it doesn't serve the same purpose as the others (hence the prefix `formatter` vs `format` for the others).

szist replied:
I get that, but what's the benefit of not storing the function here? It's not consistent like this.

paldys replied:
In this case, I would recommend to inline it as it is. It would be less confusing that way.

Btw, there is an error in this, so I will fix that too.

paldys commented on client/src/utils/format.js:
```diff
@@ -0,0 +1,17 @@
+export const dateFormat = (date) => (date ? dateFormatter.format(date, 'd.M.yyyy') : '')
+
+export const parseDate =
+  (dateString, baseDate = new Date()) => parse(dateString, 'yyyy-MM-dd', baseDate)
```
Oh.. I haven't really looked into that, I will check it. Thanks for the heads-up. :+1:

szist commented on client/src/pages/PayrollPage.js:
```diff
@@ -174,25 +136,37 @@ class OverviewPage extends Component {
+    const outFlow = _.sum(selectedAmounts.filter((v) => v < 0))
+    const inFlow = _.sum(selectedAmounts.filter((v) => v > 0))
+
+    return (<Fragment>
```
I think formatting wise Fragment should be on a newline, but I'm not sure. It would be good to add a custom vacuumlabs config to prettier then apply it here.

hleumas commented on client/src/App.js:
```diff
@@ -57,31 +58,37 @@ class App extends Component {
-            <link rel="icon" type="image/png" href={favicon16} size="16x16" />
-            <link rel="icon" type="image/png" href={favicon32} size="32x32" />
-          </Helmet>
+        <BrowserRouter>
```
Overall, you should write better commit messages.

hleumas commented on client/src/pages/PayrollPage.js:
```diff
@@ -93,7 +93,7 @@ class PayrollPage extends Component {
             item.date = parse(item.date, 'yyyy-MM-dd', new Date())
             return item
           })
-          .sort(compareAsc)
```
Is `compareAsc` function still used anywhere? If not, why don't we remove it?

hleumas commented on client/src/pages/PayrollPage.js:
```diff
@@ -89,8 +82,8 @@ class PayrollPage extends Component {
             item.id = id
-            item.created_at = parse(item.created_at, 'yyyy-MM-dd', new Date())
-            item.date = parse(item.date, 'yyyy-MM-dd', new Date())
+            item.created_at = new Date(item.created_at)
```
This certainly isn't related to "extracting formatting to a common file", is it?

hleumas commented on client/src/api.js:
```diff
@@ -0,0 +1,31 @@
+    throw new Error('server error')
+  }
+
+  const contentType = response.headers.get('content-type')
```
This commit does much more than just "extracting API calls to a separate file". Start writing usefull commit messages please.

hleumas commented on server/src/frontend.js:
```diff
@@ -7,7 +7,7 @@ export default router
 router.use(express.static(path.join(clientBuildDir)))
 
-router.all('/', (req, res, next) => {
+router.all('*', (req, res, next) => {
```
Is this really a frontend code?

hleumas commented on client/src/pages/pageStyle.js:
```diff
@@ -0,0 +1,33 @@
+import {color} from '../components/style'
```
You don't really need to call file `pageStyle` if it's in the directory `pages`. Keep it simple, don't repeat yourself.

hleumas commented on client/src/pages/pageStyle.js:
```diff
@@ -0,0 +1,33 @@
+    flexWrap: 'wrap',
+    margin: theme.spacing.unit,
+    padding: '25px 0',
+    zIndex: 100,
```
This does much more than just extracting style, you've actually changed the styles along the way. Don't do that.

hleumas commented on client/src/pages/SharesPage.js:
```diff
@@ -0,0 +1,85 @@
+  {id: 'created_at', numeric: false, isDate: true, label: 'Created at', pctWidth: 10},
+]
+
+class SharesPage extends Component {
```
I'd guess we can do much better work reusing components.

hleumas commented on client/src/components/Fetch.js:
```diff
@@ -0,0 +1,43 @@
+class Fetch extends Component {
+  state = {
+    data: [],
+    loading: true,
```
We don't need `loading`, we can simply check whether `data!=null`.

paldys commented:
Since last time, I implemented the suggestion, I rebased the branch to master and I extracted the date filtering function (`isWithinRange` to `DateFilter.js`).

If you decide to merge, please don't forget to set `po_sheet_shares_id` in `Now` to point to the correct google sheet.
